helix-term/src/commands/lsp.rs

```diff
@@ -1324,3 +1330,170 @@ pub fn select_references_to_symbol_under_cursor(cx: &mut Context) {
+            );
+
+            (
+                language_server.text_document_range_inlay_hints(doc.identifier(), range, None),
+                new_doc_inlay_hint_id,
+            )
+        }
+        _ => return None,
+    };
+
+    let callback = super::make_job_callback(
+        future?,
+        move |editor, _compositor, response: Option<Vec<lsp::InlayHint>>| {
+            // The config was modified or the window was closed while the request was in flight
+            if !editor.config().lsp.display_inlay_hints || editor.tree.try_get(view_id).is_none() {
+                return;
+            }
+
+            // Add annotations to relevant document, not the current one (it may have changed in between)
+            let doc = match editor.documents.get_mut(&doc_id) {
+                Some(doc) => doc,
+                None => return,
+            };
+
+            let mut hints = match response {
+                Some(h) if !h.is_empty() => h,
+                _ => return,
+            };
+
+            let offset_encoding = match doc.language_server() {
+                Some(ls) => ls.offset_encoding(),
+                None => return,
+            };
+
+            // Most language servers will already send them sorted but ensure this is the case to
+            // avoid errors on our end.
+            hints.sort_unstable_by_key(|inlay_hint| inlay_hint.position);
+
+            let mut type_inlay_hints = Vec::new();
+            let mut parameter_inlay_hints = Vec::new();
+            let mut other_inlay_hints = Vec::new();
+
+            let doc_text = doc.text();
+
+            for hint in hints {
+                let char_idx =
+                    match helix_lsp::util::lsp_pos_to_pos(doc_text, hint.position, offset_encoding)
+                    {
+                        Some(pos) => pos,
+                        // Skip inlay hints that have no "real" position
+                        None => continue,
+                    };
+
+                let label = match hint.label {
+                    lsp::InlayHintLabel::String(s) => s,
+                    lsp::InlayHintLabel::LabelParts(parts) => parts
+                        .into_iter()
+                        .map(|p| p.value)
+                        .collect::<Vec<_>>()
+                        .join(""),
+                };
+
+                let inlay_hints_vec = match hint.kind {
+                    Some(lsp::InlayHintKind::TYPE) => &mut type_inlay_hints,
+                    Some(lsp::InlayHintKind::PARAMETER) => &mut parameter_inlay_hints,
+                    // We can't warn on unknown kind here since LSPs are free to set it or not, for
+                    // example Rust Analyzer does not: every kind will be `None`.
+                    _ => &mut other_inlay_hints,
+                };
+
+                if let Some(true) = hint.padding_left {
+                    inlay_hints_vec.push(InlineAnnotation::new(char_idx, " "));
+                }
```

Comment:
From the LSP spec:

> Note: Padding should use the editor's background color, not the
> background color of the hint itself. That means padding can be used
> to visually align/separate an inlay hint.

That means that you need one extra layer for the the padding which uses `ui.virtual.whitespace` (the theme key normally used for spaces)

Reply:
Fixed, though I introduced `ui.virtual.inlay-hint.padding` instead, to help people have a different style for them than regular whitespaces, else it quickly becomes confusing because there appears to be an unreachable whitespace that is not an inlay hint.